src/envoy/http/authnv2/filter.cc

```diff
@@ -0,0 +1,236 @@
+    for (auto key : keys) {
+      list->push_back(key);
+    }
+  } catch (Json::Exception& e) {
+    // Not convertable to string
+  }
+  // Next, try as string array
+  try {
+    std::vector<std::string> vector = obj.getStringArray(key);
+    for (const std::string v : vector) {
+      list->push_back(v);
+    }
+  } catch (Json::Exception& e) {
+    // Not convertable to string array
+  }
+}
+
+// Helper function to set a key/value pair into Struct.
+static void setKeyValue(::google::protobuf::Struct& data,
+                        const std::string& key, const std::string& value) {
+  (*data.mutable_fields())[key].set_string_value(value);
+}
+
+// Returns true if the attribute populated to authn filter succeeds.
+bool ProcessJwt(const std::string& jwt, ProtobufWkt::Struct& authn_data) {
+  Envoy::Json::ObjectSharedPtr json_obj;
+  try {
+    json_obj = Json::Factory::loadFromString(jwt);
+    // ENVOY_LOG(debug, "{}: json object is {}", __FUNCTION__,
+    //           json_obj->asJsonString());
+  } catch (...) {
+    return false;
+  }
+
+  ProtobufWkt::Struct claim_structs;
+  auto claims = claim_structs.mutable_fields();
+  // Extract claims as string lists
+  json_obj->iterate([json_obj, claims](const std::string& key,
+                                       const Json::Object&) -> bool {
+    // In current implementation, only string/string list objects are extracted
+    std::vector<std::string> list;
+    ExtractStringList(key, *json_obj, &list);
+    for (auto s : list) {
+      (*claims)[key].mutable_list_value()->add_values()->set_string_value(s);
+    }
+    return true;
+  });
+
+  if (claims->find(kJwtAudienceKey) != claims->end()) {
+    // TODO(diemtvu): this should be send as repeated field once mixer
+    // support string_list (https://github.com/istio/istio/issues/2802) For
+    // now, just use the first value.
+    const auto& aud_values = (*claims)[kJwtAudienceKey].list_value().values();
+    if (!aud_values.empty()) {
+      setKeyValue(authn_data,
+                  istio::utils::AttributeName::kRequestAuthAudiences,
+                  aud_values[0].string_value());
+    }
+  }
+
+  // request.auth.principal
+  if (claims->find("iss") != claims->end() &&
+      claims->find("sub") != claims->end()) {
+    setKeyValue(
+        authn_data, istio::utils::AttributeName::kRequestAuthPrincipal,
+        (*claims)["iss"].list_value().values().Get(0).string_value() + "/" +
+            (*claims)["sub"].list_value().values().Get(0).string_value());
+  }
+
+  // request.auth.audiences
+  if (claims->find("azp") != claims->end()) {
+    setKeyValue(authn_data, istio::utils::AttributeName::kRequestAuthPresenter,
+                (*claims)["azp"].list_value().values().Get(0).string_value());
+  }
+
+  // request.auth.claims
+  (*(*authn_data
+          .mutable_fields())[istio::utils::AttributeName::kRequestAuthClaims]
+        .mutable_struct_value())
+      .MergeFrom(claim_structs);
```

Comment:
Is there any validation on `claim_structs`? It is originated from json string. I don't know if it is sanitized already